76-lynx-boreal/technical_specification.md

```diff
@@ -0,0 +1,458 @@
+- `GET /users/{user_id}/contexts`: List all `UploadContext` IDs available to the user
+- `POST /work-packages/{work_package_id}/contexts/{context_id}/work-order-tokens`: Create a WOT for uploading files
+  - Requires a Work Package Access Token, so the user must have already created a Work Package
+  - The request body must contain the work type and file alias or ID, depending on the
+    work type
+
+#### Claims Repository Service:
+- CRS Authentication for upload endpoints should match existing download counterparts
+- `GET /upload-access/users/{user_id}/contexts`: lists which `UploadContext`s a user can access
+- `GET /upload-access/users/{user_id}/contexts/{context_id}`: check if a user has access to a certain upload context
+- `POST /upload-access/users/{user_id}/ivas/{iva_id}/contexts/{context_id}`: grant upload access
+  - This is called by the UOS when the Data Steward grants a user upload access
+- `DELETE /upload-access/users/{user_id}/contexts/{context_id}`: revoke upload access
+
+### Payload Schemas for Events:
+
+```python
+class UploadContextState(StrEnum):
+    """The allowed states for an UploadContext instance"""
+
+    OPEN = "open"
+    LOCKED = "locked"
+    CLOSED = "closed"
+
+class UploadContext(BaseModel):
+    """A class representing an Upload Context"""
+
+    context_id: UUID4 # unique identifier for the instance
+    state: UploadContextState  # one of OPEN, LOCKED, CLOSED
+    name: str  # short meaningful name for the context
+    description: str  # used to help distinguish contexts since they aren't linked to studies
+    file_uploads: list[FileUpload]  # use list function for default_factory
+
+class FileUploadState(StrEnum):
+    """The allowed states for a FileUpload instance"""
+
+    INIT = "init"
+    COMPLETED = "completed"
+
+class FileUpload(BaseModel):
+    """A File Upload"""
+
+    upload_id: UUID4
+    state: FileUploadState  # one of INIT, COMPLETED
+    alias: str  # the submitted alias from the metadata
+    checksum: str
+```
+
+
+## Additional Implementation Details:
+
+### WOT Modifications in WPS
+The WPS needs to be able to authorize work in a compartmentalized fashion.
+Instead of adapting the existing WOT schema to work for multiple services,
+there should be specific schemas dedicated to a given action (work order).
+The existing WorkOrderToken model must be modified & renamed.
+Accordingly, new WOT schemas should be created in the WPS:
+
+```python
+DownloadFileWorkOrder:  # replaces current `WorkOrderToken` model
+   type: "download"
+   file_id: str  # note: user information is no longer included
+
+CreateUploadContextWorkOrder:
+  type: "create"
+  name: str
+
+ChangeUploadContextWorkOrder:
+   type: "lock" | "close" | "open"
+   context_id: str
+
+ViewUploadContextWorkOrder:
+   type: "view"
+   context_ids: list[str]
+
+CreateFileWorkOrder:
+  type: "create"
+  alias: str
+  context_id: str
+
+UploadFileWorkOrder:
+   type: "upload" | "close" | "delete"
+   file_id: str
```

Comment:
The classes should be ordered by issuer, i.e. first the work orders for files, issued by the WPS, and then the work orders for upload contexts ("boxes") issued by the UOS.